rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/parametric/model/ParametricSummary.java

```diff
@@ -0,0 +1,500 @@
+ */
+abstract class ParametricSummaryFactory {
+    public static final String DOCUMENTATION = "documentation";
+    public static final String DEFINITIONS = "definitions";
+    public static final String REFERENCES = "references";
+    public static final String IMPLEMENTATIONS = "implementations";
+
+    protected final SummaryConfig config;
+    protected final Executor exec;
+    protected final ColumnMaps columns;
+
+    protected ParametricSummaryFactory(SummaryConfig config, Executor exec, ColumnMaps columns) {
+        this.config = config;
+        this.exec = exec;
+        this.columns = columns;
+    }
+
+    @SuppressWarnings("deprecation") // For `MarkedString`
+    public static Either<String, MarkedString> mapValueToString(IValue v) {
+        return Either.forLeft(((IString) v).getValue());
+    }
+}
+
+class ScheduledSummaryFactory extends ParametricSummaryFactory {
+    private static final Logger logger = LogManager.getLogger(ScheduledSummaryFactory.class);
+
+    private final ScheduledCalculator calculator;
+
+    public ScheduledSummaryFactory(SummaryConfig config, Executor exec, ColumnMaps columns, ScheduledCalculator calculator) {
+        super(config, exec, columns);
+        this.calculator = calculator;
+    }
+
+    public CompletableFuture<Versioned<ParametricSummary>> createMessagesOnlySummary(
+            ISourceLocation file, CompletableFuture<Versioned<ITree>> tree) {
+        return createSummary(file, tree, MessagesOnlyScheduledSummary::new);
+    }
+
+    public CompletableFuture<Versioned<ParametricSummary>> createFullSummary(
+            ISourceLocation file, CompletableFuture<Versioned<ITree>> tree) {
+        return createSummary(file, tree, FullScheduledSummary::new);
+    }
+
+    private CompletableFuture<Versioned<ParametricSummary>> createSummary(
+            ISourceLocation file, CompletableFuture<Versioned<ITree>> tree,
+            Function<InterruptibleFuture<IConstructor>, ParametricSummary> constructor) {
+
+        return tree.thenApplyAsync(t -> {
+            logger.trace("Requesting summary calculation for: {}", file);
+            var calculation = calculator.apply(file, t.get());
+            return new Versioned<>(t.version(), constructor.apply(calculation));
+        }, exec);
+    }
+
+    public class MessagesOnlyScheduledSummary extends NullSummary {
+        private final InterruptibleFuture<List<Diagnostic>> messages;
+
+        public MessagesOnlyScheduledSummary(InterruptibleFuture<IConstructor> calculation) {
+            this.messages = extractMessages(calculation);
+        }
+
+        @Override
+        public InterruptibleFuture<List<Diagnostic>> getMessages() {
+            return messages;
+        }
+
+        @Override
+        public void invalidate() {
+            messages.interrupt();
+        }
+
+        private InterruptibleFuture<List<Diagnostic>> extractMessages(InterruptibleFuture<IConstructor> summary) {
+            return summary.thenApply(s -> {
```

Comment:
if we move this to a lazy one in the constructor, we can avoid recalculating the messages every time. or do we know it's only called once?

Reply:
Fixed